app/cdap/components/Connections/Browser/ParsingConfigModal/store.tsx

```diff
@@ -0,0 +1,222 @@
+      return {
+        ...state,
+        loadingSubcription: action.value,
+      };
+    case SET_PLUGIN_PROPERTIES_ACTION:
+      return {
+        ...state,
+        pluginProperties: action.value,
+      };
+    default:
+      throw new Error();
+  }
+};
+
+// Copied from PluginListWidget/index
+// TODO Centralize server-defined types
+interface IPlugin {
+  name: string;
+  type: string;
+  description: string;
+  className: string;
+  artifact: {
+    name: string;
+    version: string;
+    scope: string;
+  };
+}
+
+interface IConnectionSpecification {
+  relatedPlugins: IPlugin[];
+}
+
+export const performInitialLoad = (connection, sampleProperties, entity, dispatch) => {
+  const selectedProperties = {};
+  // set initial properties
+  sampleProperties.forEach((sp) => {
+    let value = null;
+    if (entity.properties[sp.name]) {
+      value = entity.properties[sp.name].value;
+    }
+    selectedProperties[sp.name] = value;
+  });
+
+  const namespace = getCurrentNamespace();
+
+  const params = {
+    context: namespace,
+    connectionId: connection,
+  };
+
+  const body = {
+    path: entity.path,
+    properties: {},
+  };
+
+  let widgetKey;
+  const widgetObs = ConnectionsApi.getSpecification(params, body).pipe(
+    switchMap((spec: IConnectionSpecification) => {
+      const sourcePlugin = spec.relatedPlugins.find((plugin) => plugin.type === 'batchsource');
+      widgetKey = `widgets.${sourcePlugin.name}-${sourcePlugin.type}`;
+      const { artifact } = sourcePlugin;
+      const widgetJsonParams = {
+        namespace,
+        artifactName: artifact.name,
+        artifactVersion: artifact.version,
+        scope: artifact.scope,
+        keys: widgetKey,
+      };
+      const pluginDetailsParams = {
+        namespace: getCurrentNamespace(),
+        artifactId: artifact.name,
+        version: artifact.version,
+        scope: artifact.scope,
+        extensionType: sourcePlugin.type,
+        pluginName: sourcePlugin.name,
+      };
+
+      return Observable.forkJoin(
+        MyPipelineApi.fetchWidgetJson(widgetJsonParams),
+        MyArtifactApi.fetchPluginDetails(pluginDetailsParams)
+      );
+    })
+  );
+
+  const loadingSubscription = widgetObs.subscribe(
+    ([widgetJson, pluginDetails]) => {
+      const sourceWidgets = JSON.parse(widgetJson[widgetKey]);
+
+      const parsingWidgets = [];
+      sourceWidgets['configuration-groups'].forEach((cfgGroup) => {
+        cfgGroup.properties.forEach((propertyWidget) => {
+          if (
+            sampleProperties.some((sampleProperty) => propertyWidget.name === sampleProperty.name)
+          ) {
+            parsingWidgets.push(propertyWidget);
+
+            // set default value
+            if (!selectedProperties[propertyWidget.name] && propertyWidget['widget-attributes']) {
+              selectedProperties[propertyWidget.name] = propertyWidget['widget-attributes'].default;
+            }
+          }
+        });
+      });
+
+      dispatch({
```

Comment:
can we combine all these into one dispatch call? `SET_INITIAL_PROPERTIES` or something like that and set all the values at once